### PR TITLE
Add improved Tensor tooltips for Python files which have been run in the interactive window

### PR DESCRIPTION
**Pull request by joyceerhl**
For #302

Extends https://github.com/microsoft/vscode-python/pull/11756 by introducing a `jupyter.variableTooltipFields` setting allowing configuration of tooltips based on type.

image

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [x] Unit tests & system/integration tests are added/updated.
-   [x] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [x] ~[`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).~

### CODE REVIEW

File: package.json
Comment (joyceerhl) [inline package.json:1786]: Ran this setting past jmew and greazer, they thought having a setting for this was fine. We may be changing which attributes we show users based on a QP anyway.

File: src/test/datascience/editor-integration/hoverProvider.vscode.test.ts
Comment (joyceerhl) [inline src/test/datascience/editor-integration/hoverProvider.vscode.test.ts:39]: Test is still failing because kernelVariables doesn't appear to know anything about our variables, will fix before merge.

File: src/client/datascience/jupyter/kernelVariables.ts
Comment (joyceerhl) [inline src/client/datascience/jupyter/kernelVariables.ts:425]: The alternative is to fetch these attributes on the original IJupyterVariable, which might help with returning a hover more quickly.

File: package.json
Comment (IanMatthewHuff) [reply package.json:1786]: It is a bit of a complex setting. Not sure how many customers would understand it based on the description. But doesn't hurt to have I suppose. Even if no one external uses it, it keeps us thinking in a more extensible way.

Comment (IanMatthewHuff) [reply package.json:1786]: Having keyed by language when the description say that it's only for python is a bit funny.

File: src/client/datascience/editor-integration/hoverProvider.ts
Comment (IanMatthewHuff) [inline src/client/datascience/editor-integration/hoverProvider.ts:77]: Just for info, why was the change needed here? Is it slower now and was getting called too often?

File: package.json
Comment (joyceerhl) [reply package.json:1786]: Yeah, that's true. I do think it should be keyed by language but I don't believe we support executing interactive cells in other languages in the IW today, so I didn't want the description to imply that other languages should work now.

Comment (DonJayamanne) [reply package.json:1786]: I personally think we shouldn't be adding this. Based on what I've learnt over the years & what other PMs have said, this isn't the best way. To quote a famous man `You don't hand out a cookie to someone & ask if they want a cookie`. 
But, not going to block this PR for this.

Comment (DonJayamanne) [reply package.json:1786]: Would we remove this property if its not used?
If 10 out of 10,000 people use this, would we still keep this property or remote it.

The draw back of adding such properties is the properties window gets overloaded with 1000s of items (as reported by plenty of PMs as well).
@jmew /cc

File: src/client/common/configSettings.ts
Comment (IanMatthewHuff) [inline src/client/common/configSettings.ts:99]: I don't know that I would sink a ton of time into this. But could we examine if we could make this a specific type, and not JSONObject? JSONObject captures like...everything I think. I would think that we could get a specific interface for the format that we expect here.

File: src/client/datascience/editor-integration/hoverProvider.ts
Comment (joyceerhl) [reply src/client/datascience/editor-integration/hoverProvider.ts:77]: The original timeout was too aggressive, so even without my changes the tooltips would show up maybe 30% of the time because of this timeout.

File: package.json
Comment (DonJayamanne) [reply package.json:1786]: Why not remove this from `package.json` & we support it.
Just as we support other settings but not advertised in `package.json`
Blocking based on the request that we remove this from here & just add support for it.

We do this today for `"jupyter.disableZMQSupport"`, its a property that is supported in settings.json but not advertised, so users don't see it.

@joyceerhl @jmew (this way we can customize it for QPs & never have to add it unless users need it).

File: .github/test_plan.md
Comment (DonJayamanne) [inline .github/test_plan.md:73]: Do we need this if we have an end to end test for this.
I think we should only add stuff that's difficult to test via code

File: package.json
Comment (joyceerhl) [reply package.json:1786]: Good idea, I'm fine with that.

**Review by DonJayamanne — CHANGES_REQUESTED** [review]
Can we remove the changes from package.json & leave the rest (this way you can still configure if required)

File: .github/test_plan.md
Comment (joyceerhl) [reply .github/test_plan.md:73]: Aren't we supposed to update the test plan with new functionality? I'm happy to remove this, as the test I have simulates this functionality pretty well. Just not sure what our policy on test plan updates is.

File: src/client/datascience/editor-integration/hoverProvider.ts
Comment (IanMatthewHuff) [inline src/client/datascience/editor-integration/hoverProvider.ts:109]: I always forget about reduce. That's a nice way to put the string together. If it doesn't have any matching attributes does it create an empty line? Just not sure if the \n goes on the end of the entry or on the start. From the docs it looks like maybe the initial value should just be ``` and the \n would be on the start of each entry, not the end.

File: src/client/datascience/editor-integration/hoverProvider.ts
Comment (IanMatthewHuff) [inline src/client/datascience/editor-integration/hoverProvider.ts:105]: I might toss on a comment here about the use of !. Had to actually look it up myself to see if this was safe. It's a bit funny since it's optional on IJupyterVariables, but this class now demands that it exists. It works since we are specifically injecting the kernelVariables, but a comment explaining that here might be nice.

File: src/client/datascience/editor-integration/hoverProvider.ts
Comment (joyceerhl) [reply src/client/datascience/editor-integration/hoverProvider.ts:109]: If it doesn't have any matching attributes we probably shouldn't return a hover result at all.

**Review by IanMatthewHuff — APPROVED** [review]

File: .github/test_plan.md
Comment (rchiodo) [reply .github/test_plan.md:73]: I don't think it hurts anything here other than it taking longer for CTI to test.

File: src/client/common/configSettings.ts
Comment (rchiodo) [reply src/client/common/configSettings.ts:99]: I believe it can just be an object.

File: src/client/datascience/jupyter/kernelVariables.ts
Comment (rchiodo) [reply src/client/datascience/jupyter/kernelVariables.ts:425]: I have another idea for this. Improve IPython such that it returns these items for tensors in the inspect request. Then we use that to get the values (this way jupyter gets improved tensor support too).

We could do that in parallel to this though.

File: src/client/datascience/jupyter/kernelVariables.ts
Comment (rchiodo) [inline src/client/datascience/jupyter/kernelVariables.ts:432]: Hardcoding python here shouldn't be necessary. Other code in this file tries to be language agnostic.

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/jupyter/kernelVariables.ts
Comment (joyceerhl) [reply src/client/datascience/jupyter/kernelVariables.ts:425]: Good idea. Would we file an issue for them and offer to submit a PR? Or just submit the PR?

Comment (rchiodo) [reply src/client/datascience/jupyter/kernelVariables.ts:425]: File an issue, then submit a PR I'd say.

Comment (joyceerhl) [reply src/client/datascience/jupyter/kernelVariables.ts:424]: Opened an issue here: https://github.com/ipython/ipython/issues/12816

**Comment by joyceerhl** [comment]
Finally got test passing 🙄 https://github.com/microsoft/vscode-jupyter/pull/4693/checks?check_run_id=1866750633#step:31:942

**Review by rchiodo — APPROVED** [review]

**Comment by joyceerhl** [comment]
@DonJayamanne I think I've addressed your feedback (package.json should not list setting).

File: src/client/datascience/jupyter/kernelVariables.ts
Comment (joyceerhl) [reply src/client/datascience/jupyter/kernelVariables.ts]: Oh wait, didn't see this. GitHub shows outdated code snippets for stuff that's been reviewed, so I think you were looking at the old version. I initially passed the language through from the hover provider request, but I much prefer this idea and it's more consistent with what the rest of kernelVariables does. Thanks for the suggestion :)

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**
